src/lib/transition_handler/validator.ml

```diff
@@ -58,13 +75,27 @@ module Make (Inputs : Inputs.S) :
            let transition_with_hash = {With_hash.hash; data= transition} in
            Deferred.return
              ( match
-                 validate_transition ~logger ~frontier transition_with_hash
+                 validate_transition ~logger ~frontier
+                   ~unprocessed_transition_cache transition_with_hash
                with
              | Ok () ->
                  Logger.info logger
                    !"accepting transition %{sexp:State_hash.t}"
                    hash ;
-                 Writer.write valid_transition_writer transition_with_hash
+                 let cached_transition :
+                     ( ( External_transition.Verified.t
+                       , State_hash.t )
+                       With_hash.t
+                     , State_hash.t )
+                     Cached.t =
+                   (* since validate_transition checks this
+                    * cache, we can expect the call to this
+                    * function to only return Ok *)
+                   Unprocessed_transition_cache.register
+                     unprocessed_transition_cache transition_with_hash
+                   |> Or_error.ok_exn
```

Comment:
it's kind of dirty to decouple the check and the unwrapping here -- can you instead register in the `validate_transition` and return a `_ Cached.t Or_error.t` instead of a `unit Or_error.t`?